src/frontend/flight/sync/utils.rs

```diff
@@ -0,0 +1,545 @@
+use datafusion_common::{DataFusionError, Result};
+use datafusion_expr::{lit, Accumulator, Expr};
+use itertools::Itertools;
+use std::collections::{HashMap, VecDeque};
+
+// Compact a set of record batches into a single one, squashing any chain of changes to a given row
+// into a single row in the output batch.
+// This means that if a row is changed multiple times, only the last change will be reflected in the
+// output batch (meaning the last NewPk and Value role columns and the last Value column where the
+// accompanying Changed field was `true`).
+pub(super) fn compact_batches(
+    sync_schema: &SyncSchema,
+    data: Vec<RecordBatch>,
+) -> Result<RecordBatch> {
+    // Concatenate all the record batches into a single one
+    let schema = data.first().unwrap().schema();
+    let batch = concat_batches(&schema, &data)?;
+
+    // Get columns, sort fields and null arrays for a particular role
+    let columns = |role: ColumnRole| -> (Vec<ArrayRef>, (Vec<SortField>, Vec<ArrayRef>)) {
+        sync_schema
+            .columns()
+            .iter()
+            .zip(batch.columns().iter())
+            .filter_map(|(col, array)| {
+                if col.role() == role {
+                    let data_type = array.data_type();
+                    Some((
+                        array.clone(),
+                        (
+                            SortField::new(data_type.clone()),
+                            new_null_array(data_type, 1),
+                        ),
+                    ))
+                } else {
+                    None
+                }
+            })
+            .unzip()
+    };
+
+    let (old_pk_cols, (sort_fields, nulls)) = columns(ColumnRole::OldPk);
+    let (new_pk_cols, _) = columns(ColumnRole::NewPk);
+
+    // NB: we must use the same row converter in order to compare fields, or else the comparison
+    // might not make sense:
+    // https://github.com/apache/arrow-rs/blob/956fe76731b80f62559e6290cfe6fb360794c3ce/arrow-row/src/lib.rs#L47-L48
+    //
+    // In our case this also necessitates that we use the same sort fields for both old and new PKs
+    // and so need to validate this during `SyncSchema` construction.
+    //
+    // If we ever need to support PK evolution this could be achieved by projecting any missing
+    // columns as nulls before constructing the rows.
+    let converter = RowConverter::new(sort_fields)?;
+    let old_pks = converter.convert_columns(&old_pk_cols)?;
+    let new_pks = converter.convert_columns(&new_pk_cols)?;
+    let null_rows = converter.convert_columns(&nulls)?;
+    let nulls = null_rows.row(0);
+
+    let (changed_binding, _) = columns(ColumnRole::Changed);
+    let changed = changed_binding
+        .iter()
+        .map(|array| {
+            array
+                .as_any()
+                .downcast_ref::<arrow::array::BooleanArray>()
+                .expect("Changed column must be boolean")
```

Comment:
Aside: do you know what happens if we panic in the Flight handler? Does the entire Seafowl process die or it's just that the gRPC caller gets an error response?

Reply:
Good question, I think it might be just the flight frontend task that dies? Need to check.

Reply:
Tried it, apparently nothing noteworthy happens apart from the client receiving an error:
```
Tonic(Status { code: Cancelled, message: "h2 protocol error: http2 error: stream error received: stream no longer needed", source: Some(tonic::transport::Error(Transport, hyper::Error(Http2, Error { kind: Reset(StreamId(1), CANCEL, Remote) }))) })
```
and seafowl logging
```
thread 'flight::client::test_basic_queries' panicked at src/frontend/flight/sql.rs:50:9:
Test handler panic
stack backtrace:
...
```
for each panic.

I interleaved a couple of calls to the handshake (in which I've added a panic) in some tests and while that particular grpc call has the error the rest finish successfully.